README.md
# ring-discord-auth

`ring-discord-auth` provides functions to verify ED-25519 signatures sent by Discord when using [webhook-based interactions](https://discord.com/developers/docs/interactions/slash-commands#receiving-an-interaction).\
Despite its name, it need not be used with ring; the library provides general functions to integrate the authentication process into any system.

This library does not provide support for Discord's OAuth2.

## Installation

You can add the library through the dependency below.

[![Clojars Project](https://img.shields.io/clojars/v/com.github.johnnyjayjay/ring-discord-auth.svg)](https://clojars.org/com.github.johnnyjayjay/ring-discord-auth)

## Usage

Below is an example of a minimal Discord app that uses the library. Here, a synchronous handler is used, but the middleware also supports asynchronous handlers.

``` clojure
(ns example.core
  (:gen-class)
  (:require [ring-discord-auth.ring :refer [wrap-authenticate]]
            [ring.middleware.json :refer [wrap-json-body wrap-json-response]]
            [ring.util.response :refer [response]]
            [org.http-kit.server :as http]))

(defn handler [{{:keys [type]} :body :as _request}]
  (response
   (case type
     1 {:type 1} ; Respond to PING with PONG
     2 {:type 4 :data {:content "Hello!"}} ; Respond to a slash command with "Hello!"
     3 {:type 6}))) ; ACK component presses but do nothing further

(def public-key "Your app's public key (can be found in the developer portal)")

(defn -main [& args]
  (http/run-server
   (-> handler
       wrap-json-response
       (wrap-json-body {:keywords? true})
       (wrap-authenticate public-key))))
```

Note that `wrap-authenticate` requires access to the raw, unmodified body and must therefore be run before `wrap-json-body`.

Or you can use the `ring-discord-auth.core/authentic?` directly.

```clojure
(ns example.core
  (:gen-class)
  (:require [ring-discord-auth.core :as discord-auth]))

(defn -main [& args]
  (let [public-key-hex "e421dceefff3a9d008b7898fcc0974813201800419d72f36d51e010d6a0acb71"
        timestamp "1625603592"
        body "this should be a json."
        signature "f31a129c4e06d93e195ea019392fc568fa7d63c9b43beb436d75f6826d5e5d36270763ee438f13ad5686ed310e8fa3253426af798927bf69cee2ff21be589109"]
    (discord-auth/authentic? signature body timestamp public-key-hex "utf8")))

;=> true
```

## License

Copyright © 2021 JohnnyJayJay, RafaelDelboni

This program and the accompanying materials are made available under the
terms of the MIT License which is available at
https://mit-license.org/.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [1.0.1]
### Fixed
- `wrap-authenticate` now accepts the same values for `public-key` as `authentic?`

## [1.0.0]
### Added
- Unit tests
- GitHub test action
- `ring-discord-auth.core/verify` as a replacement for plain signature validation in `authentic?`
- `ring-discord-auth.core/new-verifier` and `public-key->signer-verifier` for convenience with the new library
- `authentic?` accepts an `Ed25519Signer` as public key

### Changed
- `wrap-authenticate` is now in `ring-discord-auth.ring`
- Use of bouncycastle library instead of caesium (native libsodium is not required anymore)

### Removed
- Simple `authentic?` arity (replaced by `verify` function)

## 0.2.1
### Changed
- New Licence: MIT

## 0.2.0
### Added 
- Support for async ring handlers in `wrap-authenticate` middleware

## 0.1.0
### Added
- Initial project state

[1.0.0]: https://github.com/JohnnyJayJay/ring-discord-auth/tree/1.0.0
[1.0.1]: https://github.com/JohnnyJayJay/ring-discord-auth/tree/1.0.1
[Unreleased]: https://github.com/JohnnyJayJay/ring-discord-auth/tree/develop


test/ring_discord_auth/core_test.clj
(ns ring-discord-auth.core-test
  (:require [clojure.test :refer [deftest is testing]]
            [ring-discord-auth.core :as core]
            [ring-discord-auth.test-helpers :as test-helpers])
  (:import (org.bouncycastle.crypto.params Ed25519PublicKeyParameters)
           (org.bouncycastle.crypto.signers Ed25519Signer)))

(deftest public-key-verifier-test
  (let [public-key-str "e421dceefff3a9d008b7898fcc0974813201800419d72f36d51e010d6a0acb71"]
    (testing "public-key->verifier should convert all possible types or return nil"
      (is (instance? Ed25519Signer
                     (core/public-key->signer-verifier public-key-str)))

      (is (instance? Ed25519Signer
                     (core/public-key->signer-verifier (core/hex->bytes public-key-str))))

      (is (instance? Ed25519Signer
                     (core/public-key->signer-verifier (-> public-key-str
                                                           core/hex->bytes
                                                           (Ed25519PublicKeyParameters. 0)))))
      (is (instance? Ed25519Signer
                     (core/public-key->signer-verifier (-> public-key-str
                                                           core/hex->bytes
                                                           (Ed25519PublicKeyParameters. 0)
                                                           core/new-verifier))))
      (is (nil? (core/public-key->signer-verifier 1)))
      (is (nil? (core/public-key->signer-verifier nil))))))

(deftest authentic-test
  (let [key-pair (test-helpers/generate-keypair)
        signer (test-helpers/new-signer (:private key-pair))
        public-key (.getEncoded (:public key-pair))
        timestamp "1625603592"
        body "this should be a json."
        signature (->> (str timestamp body) .getBytes (test-helpers/sign signer))]
    (testing "authentic? should check signature vs public-key, timestamp and body"
      (is (= true
             (core/authentic? (test-helpers/bytes->hex signature)
                              body
                              timestamp
                              (test-helpers/bytes->hex public-key)
                              "utf8"))
          "checks with conversions.")
      (is (= true
             (core/authentic? signature
                              (.getBytes body)
                              (.getBytes timestamp)
                              public-key
                              "utf8"))
          "checks without conversions.")
      (is (= false
             (core/authentic? (test-helpers/bytes->hex signature)
                              (str body "hacks-to-fail")
                              timestamp
                              (test-helpers/bytes->hex public-key)
                              "utf8"))
          "checks with conversions.")
      (is (= false
             (core/authentic? signature
                              (.getBytes (str body "hacks-to-fail"))
                              (.getBytes timestamp)
                              public-key
                              "utf8"))
          "checks without conversions."))))


test/ring_discord_auth/ring_test.clj
(ns ring-discord-auth.ring-test
  (:require [clojure.java.io :as io]
            [clojure.test :refer [deftest is testing]]
            [ring-discord-auth.ring :as interceptor]
            [ring-discord-auth.test-helpers :as test-helpers]))

(defn build-request
  ([timestamp body signature]
   (build-request timestamp body signature :post))
  ([timestamp body signature method]
   {:request-method method
    :headers {"x-signature-ed25519" signature
              "x-signature-timestamp" timestamp}
    :body (io/input-stream (.getBytes body))
    :character-encoding "utf8"}))

(deftest verify-request-test
  (let [key-pair (test-helpers/generate-keypair)
        signer (test-helpers/new-signer (:private key-pair))
        public-key-hex (test-helpers/bytes->hex (.getEncoded (:public key-pair)))
        interceptor-fn (interceptor/wrap-authenticate identity public-key-hex)
        timestamp "1625603592"
        body "this should be a json."
        signature (->> (str timestamp body) .getBytes (test-helpers/sign signer) test-helpers/bytes->hex)]
    (testing "interceptor should check signature vs public-key, timestamp and body"
      (is (= {:request-method :post :character-encoding "utf8"}
             (-> (interceptor-fn (build-request timestamp
                                                body
                                                signature))
                 (select-keys [:request-method :character-encoding]))))
      (is (= {:status 401, :headers {"Content-Type" "text/plain", "Allow" "POST"}, :body "Signature was not authentic."}
             (-> (interceptor-fn (build-request timestamp
                                                (str body "hackedbody")
                                                signature))
                 (select-keys [:status :headers :body]))))
      (is (= {:status 400, :headers {"Content-Type" "text/plain", "Allow" "POST"}, :body "Missing body, signature or timestamp."}
             (-> (interceptor-fn (build-request nil
                                                ""
                                                nil))
                 (select-keys [:status :headers :body]))))
      (is (= {:status 405, :headers {"Content-Type" "text/plain", "Allow" "POST"}, :body "Only POST requests are allowed"}
             (-> (interceptor-fn (build-request nil
                                                ""
                                                nil
                                                :get))
                 (select-keys [:status :headers :body])))))))


test/ring_discord_auth/test_helpers.clj
(ns ring-discord-auth.test-helpers
  (:require [clojure.string :as str])
  (:import (java.security SecureRandom)
           (org.bouncycastle.crypto.generators Ed25519KeyPairGenerator)
           (org.bouncycastle.crypto.params Ed25519KeyGenerationParameters Ed25519PrivateKeyParameters Ed25519PublicKeyParameters)
           (org.bouncycastle.crypto.signers Ed25519Signer)))

(defn bytes->hex
  "Convert byte array to hex string."
  [^bytes byte-array]
  (let [hex [\0 \1 \2 \3 \4 \5 \6 \7 \8 \9 \a \b \c \d \e \f]]
    (letfn [(hexify-byte [b]
              (let [v (bit-and b 0xFF)]
                [(hex (bit-shift-right v 4)) (hex (bit-and v 0x0F))]))]
      (str/join (mapcat hexify-byte byte-array)))))

(defn generate-keypair
  "Generate Ed25519 key pair.

  Return {:private `Ed25519PrivateKeyParameters`
          :public `Ed25519PublicKeyParameters`}"
  []
  (let [random (SecureRandom.)
        kpg    (Ed25519KeyPairGenerator.)
        _ (.init kpg (Ed25519KeyGenerationParameters. random))
        key-pair (.generateKeyPair kpg)]
    {:private (cast Ed25519PrivateKeyParameters (.getPrivate key-pair))
     :public  (cast Ed25519PublicKeyParameters (.getPublic key-pair))}))

(defn new-signer
  "Return new instance of `Ed25519Signer` initialized by private key"
  [private-key]
  (let [signer (Ed25519Signer.)]
    (.init signer true private-key)
    signer))

(defn sign
  "Generate signature for msg byte array.

  Return byte array with signature."
  [^Ed25519Signer signer msg-bytes]
  (.update signer msg-bytes 0 (alength msg-bytes))
  (.generateSignature signer))


src/ring_discord_auth/core.clj
(ns ring-discord-auth.core
  (:require [clojure.java.io :as io]
            [clojure.test :refer [is]])
  (:import (java.io ByteArrayOutputStream InputStream)
           (java.nio ByteBuffer CharBuffer)
           (java.nio.charset Charset UnsupportedCharsetException IllegalCharsetNameException CharsetEncoder)
           (java.util Arrays)
           (org.bouncycastle.crypto.params Ed25519PublicKeyParameters)
           (org.bouncycastle.crypto.signers Ed25519Signer)))

(defmacro if-let-all
  "Utility-macro - like `if-let`, but with multiple bindings that are all tested."
  {:style/indent 1}
  ([bindings then]
   `(if-let-all ~bindings ~then nil))
  ([bindings then else]
   (assert (vector? bindings))
   (let [amount (count bindings)]
     (assert (= (rem amount 2) 0))
     (assert (>= amount 2))
     `(if-let [~(first bindings) ~(second bindings)]
        ~(if (> amount 2)
           `(if-let-all ~(subvec bindings 2) ~then ~else)
           then)
        ~else))))

(defmacro examples
  "Macro to generate illustrative tests for a function based on input-output pairs.

  `f` is the function to test.
  `equals` is the function used to compare the expected and actual result.
  Each example call is a sequence where the first n - 1 items are the inputs and the last item is the expected output."
  {:style/indent 1}
  [f equals & example-calls]
  `(fn []
     ~@(for [call example-calls
             :let [in (drop-last call)
                   out (last call)
                   out-ev (eval out)]]
         `(is (~equals (~f ~@in) ~out)))))

(defn hex->bytes
  "Converts the given string representing a hexadecimal number to a byte array.

  Each byte in the resulting array comes from 2 digits in the string.
  If the string cannot be converted, returns `nil`"
  {:test (examples hex->bytes Arrays/equals
           ["68c252fa74eb9b97" #_=> (byte-array [0x68 0xc2 0x52 0xfa 0x74 0xeb 0x9b 0x97])]
           ["123" #_=> (byte-array [0x12 0x03])]
           ["garbage" #_=> nil])}
  ^bytes [^String hex-str]
  (let [len (count hex-str)
        result (byte-array (quot (inc len) 2))]
    (try
      (doseq [[i hex-part] (map-indexed vector (map (partial apply str) (partition-all 2 hex-str)))]
        (aset result i (unchecked-byte (Short/parseShort hex-part 16))))
      result
      (catch NumberFormatException _ nil))))

(defn read-all-bytes
  "Reads all bytes from either an `InputStream` or a `ByteBuffer`.

  If an `InputStream` is provided, it will be consumed, but not closed.
  Returns its result as a *new* byte array."
  {:test (examples read-all-bytes Arrays/equals
           [(ByteBuffer/allocate 0) #_=> (byte-array 0)]
           [(io/input-stream (byte-array [0x12 0xab 0x4f])) #_=> (byte-array [0x12 0xab 0x4f])]
           [(ByteBuffer/wrap (byte-array [0x67 0x2a 0x4b 0x23 0x5c]) 1 2) #_=> (byte-array [0x2a 0x4b])])}
  ^bytes [input]
  (condp instance? input
    InputStream (let [bos (ByteArrayOutputStream.)]
                  (loop [next (.read ^InputStream input)]
                    (if (== next -1)
                      (.toByteArray bos)
                      (do
                        (.write bos next)
                        (recur (.read ^InputStream input))))))
    ByteBuffer (let [len (.remaining ^ByteBuffer input)
                     result (byte-array len)]
                 (.get ^ByteBuffer input result)
                 result)))

(defn encode
  "Encodes the given string to a byte array using the given charset/encoding.

  Returns `nil` if the charset is not available or if it doesn't support encoding."
  {:test (examples encode Arrays/equals
           ["Hello, world!", "utf8" #_=> (byte-array [0x48 0x65 0x6c 0x6c 0x6f 0x2c 0x20 0x77 0x6f 0x72 0x6c 0x64 0x21])]
           ["Another one", "unknown?" #_=> nil])}
  ^bytes [^String str ^String charset-name]
  (if-let-all [^Charset cs (try (Charset/forName charset-name) (catch UnsupportedCharsetException _ nil) (catch IllegalCharsetNameException _ nil))
               ^CharsetEncoder encoder (try (.newEncoder cs) (catch UnsupportedOperationException _ nil))]
    (when (.canEncode encoder str)
      (read-all-bytes (.encode encoder (CharBuffer/wrap str))))))

(defn new-verifier
  "Return new instance of `Ed25519Signer` initialized by public key."
  [public-key]
  (let [signer (Ed25519Signer.)]
    (.init signer false public-key)
    signer))

(defn verify
  "Verify signature for a message byte array.

  Takes a signature, a message and a public key `Ed25519Signer` verifier obtained by [[new-verifier]] and checks the authenticity.

  Returns `true` if the signature is valid and `false` if not."
  [^Ed25519Signer signer ^bytes msg-bytes signature]
  (.update signer msg-bytes 0 (alength msg-bytes))
  (.verifySignature signer signature))

(defn public-key->signer-verifier
  "Takes a public key as hex string, byte array, Ed25519PublicKeyParameters or Ed25519Signer.

  Returns an instance of `Ed25519Signer`."
  [public-key]
  (cond
    (string? public-key) (-> public-key hex->bytes (Ed25519PublicKeyParameters. 0) new-verifier)
    (bytes? public-key) (-> public-key (Ed25519PublicKeyParameters. 0) new-verifier)
    (instance? Ed25519PublicKeyParameters public-key) (-> public-key new-verifier)
    (instance? Ed25519Signer public-key) public-key
    :else nil))

(defn- if-str->bytes
  [body charset]
  (cond-> body
    (string? body) (encode charset)))

(defn- combine-bytes
  [^bytes timestamp-bytes ^bytes body-bytes]
  (if-let [message-bytes (byte-array (+ (alength timestamp-bytes) (alength body-bytes)))]
    (do
      (System/arraycopy timestamp-bytes 0 message-bytes 0 (alength timestamp-bytes))
      (System/arraycopy body-bytes 0 message-bytes (alength timestamp-bytes) (alength body-bytes))
      message-bytes)))

(defn authentic?
  "Checks whether a signature from a Discord interaction is authentic.

  Takes a signature (byte array or hex string), public key (byte array, hex string or verifier), body (byte array or string), timestamp (byte array or string).
  It combines timestamp and body and uses that as the message.

  Returns `true` if the message is authentic, `false` if not."
  [signature body timestamp public-key charset-name]
  (if-let-all [signature-bytes (cond-> signature (not (bytes? signature)) hex->bytes)
               public-key-signer (public-key->signer-verifier public-key)
               body-bytes (if-str->bytes body charset-name)
               timestamp-bytes (if-str->bytes timestamp charset-name)
               message-bytes (combine-bytes timestamp-bytes body-bytes)]
    (verify public-key-signer message-bytes signature-bytes)
    false))


src/ring_discord_auth/validation.clj
(ns ring-discord-auth.validation
  (:require [clojure.java.io :as io]
            [clojure.string :as str]
            [clojure.test :refer [is]])
  (:import (java.io ByteArrayOutputStream InputStream)
           (java.nio ByteBuffer CharBuffer)
           (java.nio.charset Charset UnsupportedCharsetException IllegalCharsetNameException CharsetEncoder)
           (java.security SecureRandom)
           (java.util Arrays)
           (org.bouncycastle.crypto.generators Ed25519KeyPairGenerator)
           (org.bouncycastle.crypto.params Ed25519KeyGenerationParameters Ed25519PrivateKeyParameters Ed25519PublicKeyParameters)
           (org.bouncycastle.crypto.signers Ed25519Signer)))

(defmacro examples
  "Macro to generate illustrative tests for a function based on input-output pairs.

  `f` is the function to test.
  `equals` is the function used to compare the expected and actual result.
  Each example call is a sequence where the first n - 1 items are the inputs and the last item is the expected output."
  {:style/indent 1}
  [f equals & example-calls]
  `(fn []
     ~@(for [call example-calls
             :let [in (drop-last call)
                   out (last call)
                   out-ev (eval out)]]
         `(is (~equals (~f ~@in) ~out)))))

(defmacro if-let-all
  "Utility-macro - like `if-let`, but with multiple bindings that are all tested."
  {:style/indent 1}
  ([bindings then]
   `(if-let-all ~bindings ~then nil))
  ([bindings then else]
   (assert (vector? bindings))
   (let [amount (count bindings)]
     (assert (= (rem amount 2) 0))
     (assert (>= amount 2))
     `(if-let [~(first bindings) ~(second bindings)]
        ~(if (> amount 2)
           `(if-let-all ~(subvec bindings 2) ~then ~else)
           then)
        ~else))))

(defn read-all-bytes
  "Reads all bytes from either an `InputStream` or a `ByteBuffer`.

  If an `InputStream` is provided, it will be consumed, but not closed.
  Returns its result as a *new* byte array."
  {:test (examples read-all-bytes Arrays/equals
                   [(ByteBuffer/allocate 0) #_=> (byte-array 0)]
                   [(io/input-stream (byte-array [0x12 0xab 0x4f])) #_=> (byte-array [0x12 0xab 0x4f])]
                   [(ByteBuffer/wrap (byte-array [0x67 0x2a 0x4b 0x23 0x5c]) 1 2) #_=> (byte-array [0x2a 0x4b])])}
  ^bytes [input]
  (condp instance? input
    InputStream (let [bos (ByteArrayOutputStream.)]
                  (loop [next (.read ^InputStream input)]
                    (if (== next -1)
                      (.toByteArray bos)
                      (do
                        (.write bos next)
                        (recur (.read ^InputStream input))))))
    ByteBuffer (let [len (.remaining ^ByteBuffer input)
                     result (byte-array len)]
                 (.get ^ByteBuffer input result)
                 result)))

(defn encode
  "Encodes the given string to a byte array using the given charset/encoding.

  Returns `nil` if the charset is not available or if it doesn't support encoding."
  {:test (examples encode Arrays/equals
                   ["Hello, world!", "utf8" #_=> (byte-array [0x48 0x65 0x6c 0x6c 0x6f 0x2c 0x20 0x77 0x6f 0x72 0x6c 0x64 0x21])]
                   ["Another one", "unknown?" #_=> nil])}
  ^bytes [^String str ^String charset-name]
  (if-let-all [^Charset cs (try (Charset/forName charset-name) (catch UnsupportedCharsetException _ nil) (catch IllegalCharsetNameException _ nil))
               ^CharsetEncoder encoder (try (.newEncoder cs) (catch UnsupportedOperationException _ nil))]
              (when (.canEncode encoder str)
                (read-all-bytes (.encode encoder (CharBuffer/wrap str))))))

(defn hex->bytes
  "Converts the given string representing a hexadecimal number to a byte array.
  Each byte in the resulting array comes from 2 digits in the string.

  If the string cannot be converted, returns `nil`"
  ^bytes [^String hex-str]
  (let [len (count hex-str)
        result (byte-array (quot (inc len) 2))]
    (try
      (doseq [[i hex-part] (map-indexed vector (map (partial apply str) (partition-all 2 hex-str)))]
        (aset result i (unchecked-byte (Short/parseShort hex-part 16))))
      result
      (catch NumberFormatException _ nil))))

(defn bytes->hex
  "Convert byte array to hex string."
  [^bytes byte-array]
  (let [hex [\0 \1 \2 \3 \4 \5 \6 \7 \8 \9 \a \b \c \d \e \f]]
    (letfn [(hexify-byte [b]
              (let [v (bit-and b 0xFF)]
                [(hex (bit-shift-right v 4)) (hex (bit-and v 0x0F))]))]
      (str/join (mapcat hexify-byte byte-array)))))

(defn generate-keypair
  "Generate Ed25519 key pair.

  Return {:private `Ed25519PrivateKeyParameters`
          :public `Ed25519PublicKeyParameters`}"
  []
  (let [random (SecureRandom.)
        kpg    (Ed25519KeyPairGenerator.)
        _ (.init kpg (Ed25519KeyGenerationParameters. random))
        key-pair (.generateKeyPair kpg)]
    {:private (cast Ed25519PrivateKeyParameters (.getPrivate key-pair))
     :public  (cast Ed25519PublicKeyParameters (.getPublic key-pair))}))

(defn new-signer
  "Return new instance of `Ed25519Signer` initialized by private key"
  [private-key]
  (let [signer (Ed25519Signer.)]
    (.init signer true private-key)
    signer))

(defn sign
  "Generate signature for msg byte array.

  Return byte array with signature."
  [^Ed25519Signer signer msg-bytes]
  (.update signer msg-bytes 0 (alength msg-bytes))
  (.generateSignature signer))

(defn new-verifier
  "Return new instance of `Ed25519Signer` initialized by public key."
  [public-key]
  (let [signer (Ed25519Signer.)]
    (.init signer false public-key)
    signer))

(defn verify
  "Verify signature for msg byte array.

  Returns `true` if valid signature and `false` if not."
  [^Ed25519Signer signer msg-bytes signature]
  (.update signer msg-bytes 0 (alength msg-bytes))
  (.verifySignature signer signature))

(defn verify-request
  "Verify discord payload with app public-key, request body, signature and timestamp headers

  Returns `true` if the message is authentic, `false` if not."
  [public-key timestamp body signature]
  (verify (new-verifier (Ed25519PublicKeyParameters. (hex->bytes public-key) 0))
          (.getBytes (str timestamp body) "utf8")
          (hex->bytes signature)))

(defn authentic?
  "Checks whether a signature is authentic, given a message and a public key.

  This function has 2 arities: a general purpose one that simply delegates to a crypto library and one tailored to what Discord provides.

  The general purpose arity takes a signature, a message and a public key (all of which must be byte arrays) and checks the authenticity.
  The other arity takes a signature (byte array or hex string), public key (byte array or hex string), body (byte array or string), timestamp (byte array or string).

  It combines timestamp and body and uses that as the message.

  Returns `true` if the message is authentic, `false` if not."
  ([signature-bytes message-bytes public-key-bytes]
   (verify (new-verifier (Ed25519PublicKeyParameters. public-key-bytes 0))
            message-bytes
            signature-bytes))
  ([signature body timestamp public-key charset-name]
   (if-let-all [signature-bytes (cond-> signature (not (bytes? signature)) hex->bytes)
                public-key-bytes (cond-> public-key (not (bytes? public-key)) hex->bytes)
                ^bytes body-bytes (cond-> body (string? body) (encode charset-name))
                ^bytes timestamp-bytes (cond-> timestamp (string? timestamp) (encode charset-name))
                message-bytes (byte-array (+ (alength timestamp-bytes) (alength body-bytes)))]
     (do
       (System/arraycopy timestamp-bytes 0 message-bytes 0 (alength timestamp-bytes))
       (System/arraycopy body-bytes 0 message-bytes (alength timestamp-bytes) (alength body-bytes))
       (authentic? signature-bytes message-bytes public-key-bytes))
     false)))


src/ring_discord_auth/ring.clj
(ns ring-discord-auth.ring
  "Core namespace containing functions to handle bytes, encodings and Discord authentication"
  (:require [ring-discord-auth.core :as core :refer [if-let-all]])
  (:import (java.io ByteArrayInputStream)))

(def signature-header "x-signature-ed25519")
(def timestamp-header "x-signature-timestamp")
(def default-charset "utf8")

(def default-headers {"Content-Type" "text/plain"
                      "Allow" "POST"})

(defn wrap-authenticate
  "Ring middleware to authenticate incoming requests according to the Discord specification.

  This means:
  - If request method is not POST, respond with status 405 (Method Not Allowed)
  - Get body from the request as well as [[signature-header]] and [[timestamp-header]] from the headers
  - If any of the above is not present, respond with status 400 (Bad Request)
  - Check the parameters for authenticity as defined by Discord
    - If authentic, delegate to the `handler` with a restored body
    - If not authentic, respond with status 401 (Unauthorized)

  The `public-key` is the public key of the corresponding Discord app. It may be given as something that is accepted by [[core/public-key->signer-verifier]].
  This middleware must be in the hierarchy **before** the body is processed.

  This middleware supports both synchronous and asynchronous handlers."
  [handler public-key]
  (let [public-key (core/public-key->signer-verifier public-key)]
    (fn
      ([request respond raise]
       (let [validator (wrap-authenticate identity public-key)
             result (validator request)]
         (if (:status result)
           (respond result)
           (handler result respond raise))))
      ([{:keys [body character-encoding request-method]
         {signature signature-header timestamp timestamp-header} :headers
         :or {character-encoding default-charset}
         :as request}]
       (if (= request-method :post)
         (if-let-all [sig-bytes (some-> signature core/hex->bytes)
                      time-bytes (some-> timestamp (core/encode character-encoding))
                      body-bytes (some-> body core/read-all-bytes)]
           (if (core/authentic? sig-bytes body-bytes time-bytes public-key character-encoding)
             (handler (assoc request :body (ByteArrayInputStream. body-bytes)))
             {:status 401 :headers default-headers :body "Signature was not authentic."})
           {:status 400 :headers default-headers :body "Missing body, signature or timestamp."})
         {:status 405 :headers default-headers :body "Only POST requests are allowed"})))))
